core/backend/src/IModelDb.ts

```diff
@@ -1632,7 +1632,9 @@ export namespace IModelDb { // eslint-disable-line no-redeclare
       const iModel = this._iModel;
       const jsClass = iModel.getJsClass<typeof Element>(elProps.classFullName) as any; // "as any" so we can call the protected methods
       jsClass.onInsert(elProps, iModel);
-      const val = iModel.nativeDb.insertElement(elProps);
+
+      const json = elProps instanceof Element ? elProps.toJSON() : elProps;
+      const val = iModel.nativeDb.insertElement(json);

```

Comment:
really, we should update the id in `json` and pass it to `onInserted` below

Reply:
My concern was that onInsert/ed and onUpdate/d might want to mutate the input.
All of the implementations in imodeljs repo just pull out Ids for concurrency control.

Reply:
but they're declared to take an `ElementProps`. They'll potentially have the same problem as insert/update.

Reply:
Shouldn't we correct the mistake where `SubCategoryAppearance` and `SubCategoryAppearance.Props` have different names for the same property? That would not have been intentional. Seems like we could support both names and deprecate one of them...

Reply:
We have several cases where the persistent JSON names do not match the class property names, whether intentional or not. The mistake is in treating Entity and EntityProps as interchangeable.

Reply:
@kabentley made this change and made all the class handlers take `Readonly` props because they are supposed to only react to insert/update/delete, not mutate.

Reply:
The current behavior of `insertElement` is to update the input props.id property. I am sure that things will break if that is changed.

Reply:
it does still update the input value's id property.

Reply:
I still think that trivial differences in names between Elements and ElementProps are confusing and should be corrected. Consistency is good.

Reply:
I agree, it was always a mistake for Entity to "implement" EntityProps. I don't know what we can do without a major version change.

Reply:
BTW, our `ecschema2ts` tool that developers use to generate TypeScript wrapper code automatically adds the `implements *Props` to the classes it generates. If we decide that was a mistake, we would have to correct the `ecschema2ts` tool also.